storage/index.md

```diff
@@ -70,38 +70,117 @@ PostgreSQL have some multi-modality capability.]
+GlareDB's data sources can be classified in a few groups:
+
+- structured and semi-structured file formats. There are many
+  different interchange formats, record encoding, and archival formats
+  that allow us to write data to files, and share them between
+  programs. JSON, Parquet, CSV, BSON, and Excel are just encoded raw
+  data formats that GlareDB can interpret and present as a table. Each
+  format has different properties and can provide different benefits
+  to users. Some are easier to write and more widely supported as
+  output formats, some are easier or more efficient to read
+  programtiacally, and some are more transparent to read for humans.
+
+- external database engines and services: GlareDB provides access to
+  other database systems--like PostgreSQL, MySQL, MongoDB, SQLite,
+  Snowflake and BigQuery, to enable joining data across data sources,
+  or migrating data between formats and systems. Most of the time
+  GlareDB is able to push operations _into_ the underlying database
+  engine, which means these queries can take advantage of indexes in
+  the underlying data and reduce the amount of data it needs to
+  consider.
+
+- storage engines, like Lance, DeltaLake and Iceberg, provide
+  transactional and consistently properties that file formats cannot
+  without the overhead of a process. At the same time, unlike extenral
+  services and engines, these formats don't require a controling
+  process and can work with all of the data stored to an object store
+  like S3 or GCS. They do this by defining a protocol for writing and
+  reading files that ensure that data is consistent and that writes
+  are safe.
+
+There are exceptions and edge cases: SQLite is somewhere between a
+database and a file format, and the Parquet format has metadata that
+makes it possible to push down some operations more like an external
+database, while the implementations of Delta Lake are built upon
+Parquet and JSON. However, this classification system may be useful
+in understanding the limitations and features of a specific data
+system.
 
 ### Materialization: Roll Ups and Duplication
 
-As software engineers, "duplication" is often 
+For data sets that are changing often, having multiple copies of the
+data can be very expensive: in addition to the extra space require,
+you have to update many copies of the data any time the data changes,
+and the coordination can become quite complex. However, for historic
+data, or data that doesn't change after it's written, storing the data
+in more than one form can support different queries, or save the data
+in a grouped order or pre-compute and save various aggregates or sums
+as needed. These saved roll ups are duplicated data, but because the
+input data doesn't change, and the storage costs may be minimal, you
+can save a lot of compute time by saving (or materializing) these
+results. These kinds of transformations are always optimizations in
+service of your use case.
```

Comment:
I'm having trouble parsing this sentence and think it might need to be split up:

> However, for historic
data, or data that don't change after they're written, storing the data
in more than one form can support different queries, or save the data
in a grouped order or pre-compute and save various aggregates or sums
as needed. 

```suggestion
For data sets that change often, having multiple copies of the
data can be very expensive: in addition to the extra space required,
you have to update many copies of the data any time the data change,
and the coordination can become quite complex. However, for historic
data, or data that don't change after they're written, storing the data
in more than one form can support different queries, or save the data
in a grouped order or pre-compute and save various aggregates or sums
as needed. These saved roll-ups are duplicated data, but because the
input data don't change, and the storage costs may be minimal, you
can save a lot of compute time by saving (or materializing) these
results. These kinds of transformations are always optimizations in
service of your use case.
```